kernel/kernel-api/src/main/java/io/delta/kernel/internal/rowtracking/RowTracking.java

```diff
@@ -0,0 +1,150 @@
+        final AtomicLong currRowIdHighWatermark = new AtomicLong(readRowIdHighWaterMark(snapshot));
+        return dataActions
+            .iterator()
+            .map(
+                row -> {
+                  // Non-AddFile actions are returned unchanged
+                  if (row.isNullAt(ADD_FILE_ORDINAL)) {
+                    return row;
+                  }
+
+                  AddFile addFile = new AddFile(row.getStruct(ADD_FILE_ORDINAL));
+
+                  // Assign base row ID if missing
+                  if (!addFile.getBaseRowId().isPresent()) {
+                    final long numRecords = getNumRecordsOrThrow(addFile);
+                    addFile = addFile.withNewBaseRowId(currRowIdHighWatermark.get() + 1L);
+                    currRowIdHighWatermark.addAndGet(numRecords);
+                  }
+
+                  // Assign default row commit version if missing
+                  if (!addFile.getDefaultRowCommitVersion().isPresent()) {
+                    addFile = addFile.withNewDefaultRowCommitVersion(commitVersion);
+                  }
+
+                  return SingleAction.createAddFileSingleAction(addFile.toRow());
+                });
+      }
+    };
+  }
+
+  /**
+   * Returns a {@link DomainMetadata} action if the row ID high watermark has changed due to newly
+   * processed {@link AddFile} actions. This method should be called only when the 'rowTracking'
+   * feature is supported.
+   *
+   * @param snapshot the current snapshot of the table
+   * @param dataActions the iterable of data actions that may update the high watermark
+   */
+  public static Optional<DomainMetadata> createNewHighWaterMarkIfNeeded(
+      SnapshotImpl snapshot, CloseableIterable<Row> dataActions) {
+    checkArgument(
+        TableFeatures.isRowTrackingSupported(snapshot.getProtocol()),
+        "Row ID high watermark is updated only when feature 'rowTracking' is supported.");
+
+    final long prevRowIdHighWatermark = readRowIdHighWaterMark(snapshot);
+    // Use an AtomicLong to allow for updating the high watermark in the lambda
+    final AtomicLong newRowIdHighWatermark = new AtomicLong(prevRowIdHighWatermark);
+
+    dataActions.forEach(
+        row -> {
+          if (!row.isNullAt(ADD_FILE_ORDINAL)) {
+            AddFile addFile = new AddFile(row.getStruct(ADD_FILE_ORDINAL));
+            if (!addFile.getBaseRowId().isPresent()) {
+              newRowIdHighWatermark.addAndGet(getNumRecordsOrThrow(addFile));
+            }
+          }
+        });
+
+    return (newRowIdHighWatermark.get() != prevRowIdHighWatermark)
+        ? Optional.of(new RowTrackingMetadataDomain(newRowIdHighWatermark.get()).toDomainMetadata())
+        : Optional.empty();
+  }
+
+  /**
+   * Reads the current row ID high watermark from the snapshot, or returns a default value if
+   * missing.
+   */
+  public static long readRowIdHighWaterMark(SnapshotImpl snapshot) {
```

Comment:
does this need to be public? I can't see any other usages besides from within this class.